plugins/inventory/k8s.py

```diff
@@ -171,35 +177,43 @@ def setup(self, config_data, cache, cache_key):
             self.fetch_objects(connections)
 
     def fetch_objects(self, connections):
+        if not connections:
+            connections = [{}]
+
+        if not isinstance(connections, list):
+            raise K8sInventoryException("Expecting connections to be a list.")
+
+        for connection in connections:
+            if not isinstance(connection, dict):
+                raise K8sInventoryException("Expecting connection to be a dictionary.")
+            client = self.get_api_client(**connection)
+            name = connection.get('name', self.get_default_host_name(client.configuration.host))
+            if connection.get('namespaces'):
+                namespaces = connection['namespaces']
+            else:
+                namespaces = self.get_available_namespaces(client)
 
-        if connections:
-            if not isinstance(connections, list):
-                raise K8sInventoryException("Expecting connections to be a list.")
-
-            for connection in connections:
-                if not isinstance(connection, dict):
-                    raise K8sInventoryException("Expecting connection to be a dictionary.")
-                client = self.get_api_client(**connection)
-                name = connection.get('name', self.get_default_host_name(client.configuration.host))
-                if connection.get('namespaces'):
-                    namespaces = connection['namespaces']
-                else:
-                    namespaces = self.get_available_namespaces(client)
-                for namespace in namespaces:
-                    self.get_pods_for_namespace(client, name, namespace)
-                    self.get_services_for_namespace(client, name, namespace)
-        else:
-            client = self.get_api_client()
-            name = self.get_default_host_name(client.configuration.host)
-            namespaces = self.get_available_namespaces(client)
+            sanitized_name = self.sanitize(name)
+            self.inventory.add_group(sanitized_name)
             for namespace in namespaces:
-                self.get_pods_for_namespace(client, name, namespace)
-                self.get_services_for_namespace(client, name, namespace)
+                namespace_group = self.sanitize('namespace_{0}'.format(namespace))
+
+                self.inventory.add_group(namespace_group)
+                self.inventory.add_child(sanitized_name, namespace_group)
+
+                self.get_pods_for_namespace(client, name, namespace, namespace_group)
+                self.get_pods_from_parents(client, name, namespace, namespace_group)
 
     @staticmethod
     def get_default_host_name(host):
         return host.replace('https://', '').replace('http://', '').replace('.', '-').replace(':', '_')
 
+    @staticmethod
+    def sanitize(name):
```

Comment:
There is a function called [`_sanitize_group_name`](https://github.com/ansible/ansible/blob/48f12c14e9ff2b04c49cf6f90a65f4a9ecbb7fc7/lib/ansible/plugins/inventory/__init__.py#L153). Can we use this?

Reply:
awesome, yeah we can definitely use that